src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -710,6 +708,53 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             }
 
             return UpdateExpression(builder, node.Update(left, right, node.IsRef, node.Type));
+
+            BoundExpression fieldWithSpilledReceiver(BoundFieldAccess field, ref BoundSpillSequenceBuilder leftBuilder, bool isAssignmentTarget)
+            {
+                BoundExpression receiver;
+                var generateDummyFieldAccess = false;
+                if (field.FieldSymbol.ContainingType.IsReferenceType)
+                {
+                    // a reference type can always live across await so Spill using leftBuilder
+                    receiver = Spill(leftBuilder, VisitExpression(ref leftBuilder, field.ReceiverOpt));
+
+                    // dummy field access to trigger NRE
+                    // a.b = c will trigger a NRE if a is null on assignment,
+                    // but a.b.c = d will trigger a NRE if a is null before evaluating d
+                    // so check whether we assign to the field directly
+                    generateDummyFieldAccess = !isAssignmentTarget;
+                }
+                else if (field.ReceiverOpt is BoundArrayAccess arrayAccess)
+                {
+                    // an arrayAccess returns a ref so can only be called after the await, but spill expression and indices
+                    var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                    expression = Spill(leftBuilder, expression, RefKind.None);
+                    var indices = this.VisitExpressionList(ref leftBuilder, arrayAccess.Indices, forceSpill: true);
+                    receiver = arrayAccess.Update(expression, indices, arrayAccess.Type);
+                    // dummy array access to trigger IndexOutRangeException or NRE
+                    // we only need this if the array access is a receiver since
+                    // a[0] = b triggers a NRE/IORE on assignment
```

Comment:
sorry if this is obvious but what is IORE short for?

Reply:
Index Out of Range Exception